consensus/core/src/block_manager.rs

```diff
@@ -114,69 +114,15 @@ impl BlockManager {
-                    assert_eq!(b.ancestors().len(), ancestors.len());
-                    let mut ancestor_blocks = vec![];
-                    'ancestor: for (ancestor_ref, found) in
-                        b.ancestors().iter().zip(ancestors.into_iter())
-                    {
-                        if let Some(found_block) = found {
-                            // This invariant should be guaranteed by DagState.
-                            assert_eq!(ancestor_ref, &found_block.reference());
-                            ancestor_blocks.push(found_block);
-                            continue 'ancestor;
-                        }
-                        // blocks_to_accept have not been added to DagState yet, but they
-                        // can appear in ancestors.
-                        if blocks_to_accept.contains_key(ancestor_ref) {
-                            ancestor_blocks.push(blocks_to_accept[ancestor_ref].clone());
-                            continue 'ancestor;
-                        }
-                        // If an ancestor is already rejected, reject this block as well.
-                        if blocks_to_reject.contains_key(ancestor_ref) {
-                            blocks_to_reject.insert(b.reference(), b);
-                            continue 'block;
-                        }
-                        panic!("Unsuspended block {:?} has a missing ancestor! Ancestor not found in DagState: {:?}", b, ancestor_ref);
-                    }
-                    if let Err(e) = self.block_verifier.check_ancestors(&b, &ancestor_blocks) {
-                        warn!("Block {:?} failed to verify ancestors: {}", b, e);
-                        blocks_to_reject.insert(b.reference(), b);
-                    } else {
-                        blocks_to_accept.insert(b.reference(), b);
-                    }
-                }
-            }
-            for (block_ref, block) in blocks_to_reject {
-                self.context
-                    .metrics
-                    .node_metrics
-                    .invalid_blocks
-                    .with_label_values(&[&block_ref.author.to_string(), "accept_block"])
-                    .inc();
-                warn!("Invalid block {:?} is rejected", block);
-            }
-
-            // TODO: report blocks_to_reject to peers.
-
-            // Insert the accepted blocks into DAG state so future blocks including them as
-            // ancestors do not get suspended.
-            let blocks_to_accept: Vec<_> = blocks_to_accept.into_values().collect();
-            self.dag_state
-                .write()
-                .accept_blocks(blocks_to_accept.clone());
+            let unsuspended_blocks = self.try_unsuspend_children_blocks(block.reference());
 
+            // Verify block timestamps
+            let blocks_to_accept = self
+                .verify_block_timestamps_and_accept(iter::once(block).chain(unsuspended_blocks));
```

Comment:
Practically most of the previous code has been extracted to the `verify_block_timestamps_and_accept` method. This won't be needed once we refactor to the new timestamp approach